File: gtci/04-merge-intervals/cpp/4-minimum-meeting-rooms.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

class Meeting {
   public:
    int start = 0;
    int end = 0;

    Meeting(int start, int end) {
        this->start = start;
        this->end = end;
    }
};

class Solution {
   public:
    int findMinimumMeetingRooms(vector<Meeting>&& meetings) {
        if (meetings.size() < 1) return static_cast<int>(meetings.size());

        std::sort(meetings.begin(), meetings.end(),
                  [](Meeting a, Meeting b) { return a.start < b.start; });

        int minRooms = 0;

        auto meetingCmp = [](const Meeting& a, const Meeting& b) {
            return a.end > b.end;
        };
        std::priority_queue<Meeting, std::vector<Meeting>, decltype(meetingCmp)>
            minHeap(meetingCmp);

        for (Meeting& meeting : meetings) {
            while (!minHeap.empty() && minHeap.top().end <= meeting.start)
                minHeap.pop();

            minHeap.push(meeting);

            minRooms = std::max(minRooms, static_cast<int>(minHeap.size()));
        }

        return minRooms;
    }
};

int main() {
    Solution sol;

    assert(sol.findMinimumMeetingRooms(
               {Meeting(1, 4), Meeting(2, 5), Meeting(7, 9)}) == 2);
    assert(sol.findMinimumMeetingRooms(
               {Meeting(6, 7), Meeting(2, 4), Meeting(8, 12)}) == 1);
    assert(sol.findMinimumMeetingRooms(
               {Meeting(1, 4), Meeting(2, 3), Meeting(3, 6)}) == 2);
    assert(sol.findMinimumMeetingRooms({Meeting(4, 5), Meeting(2, 3),
                                        Meeting(2, 4), Meeting(3, 5)}) == 2);

    cout << "All test cases passed." << endl;

    return 0;
}

File: gtci/04-merge-intervals/python3/4-minimum-meeting-rooms.py
import heapq


class Interval:
    def __init__(self, start, end):
        self.start = start
        self.end = end

    def __lt__(self, other):
        return self.start < other.start


class Solution:
    def findMinimumMeetingRooms(self, meetings):
        if len(meetings) < 1:
            return len(meetings)

        min_heap = []
        min_rooms = 0

        meetings.sort(key=lambda x: x.start)

        for meeting in meetings:
            while len(min_heap) > 0 and min_heap[0][1].end <= meeting.start:
                heapq.heappop(min_heap)

            heapq.heappush(min_heap, (meeting.end, meeting))

            min_rooms = max(min_rooms, len(min_heap))

        return min_rooms


if __name__ == "__main__":
    sol = Solution()

    assert (
        sol.findMinimumMeetingRooms([Interval(1, 4), Interval(2, 5), Interval(7, 9)])
        == 2
    )
    assert (
        sol.findMinimumMeetingRooms([Interval(6, 7), Interval(2, 4), Interval(8, 12)])
        == 1
    )
    assert (
        sol.findMinimumMeetingRooms([Interval(1, 4), Interval(2, 3), Interval(3, 6)])
        == 2
    )
    assert (
        sol.findMinimumMeetingRooms(
            [Interval(4, 5), Interval(2, 3), Interval(2, 4), Interval(3, 5)]
        )
        == 2
    )

    print("All test cases passed.")
